### PR TITLE
[ASR] Multichannel mask estimator with flex number of channels

### PR DESCRIPTION
**Pull request by anteju**
# What does this PR do ?

This PR adds a multichannel mask estimator which can process flexible number of channels.
It also fixes some bugs related to PTL 2.0 and librosa version.

**Collection**: ASR

# Changelog 

- added example config `beamforming_flex_channels.yaml`
- add `max_utts` `process_audio.py`
- updated `librosa.get_duration` to use `path`, since `filename` is deprecated
- bumped `librosa` requirement to `0.10.0` in `requirements_asr.py`
- added `MaskEstimatorFlexChannels` in `audio_modules.py`
- added various multichannel modules in `multichannel_modules.py`, including `ParametricMultichannelWienerFiler` and `ReferenceChannelEstimatorSNR`
- added unit tests for multichannel modules

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/enhancement_models.py [hunk 1]
```diff
@@ -61,14 +61,24 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         self.decoder = EncMaskDecAudioToAudioModel.from_config_dict(self._cfg.decoder)
 
         if 'mixture_consistency' in self._cfg:
+            logging.debug('Using mixture consistency')
```
Comment (titu1994) [inline nemo/collections/asr/models/enhancement_models.py:64]: Are the debug statements necessary ?

File: nemo/collections/asr/models/enhancement_models.py [hunk 2]
```diff
@@ -61,14 +61,24 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
         else:
+            logging.debug('Mixture consistency not used')
             self.mixture_consistency = None
 
         # Future enhancement:
         # If subclasses need to modify the config before calling super()
         # Check ASRBPE* classes do with their mixin
 
+        # Setup augmentation
+        if hasattr(self._cfg, 'channel_augment') and self._cfg.channel_augment is not None:
```
Comment (titu1994) [inline nemo/collections/asr/models/enhancement_models.py:75]: Inside model, after constructor try to use self.cfg directly. No need for _cfg

File: nemo/collections/asr/models/enhancement_models.py [hunk 3]
```diff
@@ -397,17 +407,25 @@ def training_step(self, batch, batch_idx):
...
+        # Prepare logs
         tensorboard_logs = {
-            'train_loss': loss_value,
+            'train_loss': loss,
             'learning_rate': self._optimizer.param_groups[0]['lr'],
             'global_step': torch.tensor(self.trainer.global_step, dtype=torch.float32),
         }
 
-        return {'loss': loss_value, 'log': tensorboard_logs}
+        self.log_dict(tensorboard_logs, on_step=True, sync_dist=True)
```
Comment (titu1994) [inline nemo/collections/asr/models/enhancement_models.py:427]: So you need to sync every step ? This is training metrics right ?

File: nemo/collections/asr/modules/audio_modules.py [hunk 4]
```diff
@@ -86,6 +91,19 @@ def __init__(
             self._num_features = num_subbands
             self._num_channels = num_input_channels if self.mag_reduction is None else 1
 
+        self.eps = eps
+
+        logging.debug('Initialized %s with', self.__class__.__name__)
```
Comment (titu1994) [inline nemo/collections/asr/modules/audio_modules.py:96]: Will the debug statements be kept for later use ?

File: nemo/collections/asr/modules/audio_modules.py [hunk 5]
```diff
@@ -122,6 +140,98 @@ def num_channels(self) -> int:
...
+        """Calculate mean across time and channel dimensions.
+
+        Args:
+            input: tensor with shape (B, C, F, T)
+            input_length: tensor with shape (B,)
+
+        Returns:
+            Mean of `input` calculated across time and channel dimension
+            with shape (B, 1, F, 1)
+        """
```
Comment (titu1994) [inline nemo/collections/asr/modules/audio_modules.py:154]: Probably good to assert or check the input shape if you depend on specific layout of input

File: nemo/collections/asr/modules/audio_modules.py [hunk 6]
```diff
@@ -12,35 +12,36 @@
...
-    import torchaudio
-
-    HAVE_TORCHAUDIO = True
-except ModuleNotFoundError:
-    HAVE_TORCHAUDIO = False
-
-
 __all__ = [
     'MaskEstimatorRNN',
+    'MaskEstimatorFlexChannels',
```
Comment (titu1994) [inline nemo/collections/asr/modules/audio_modules.py:38]: Are there unittests for this?

File: nemo/collections/asr/modules/audio_modules.py [hunk 7]
```diff
@@ -342,6 +462,245 @@ def forward(self, input: torch.Tensor, input_length: torch.Tensor) -> Tuple[torc
...
+        logging.debug('Channel reduction will be applied after block %d', self.channel_reduction_position)
+
+        # Prepare processing blocks
+        self.channel_blocks = torch.nn.ModuleList()
+        self.temporal_blocks = torch.nn.ModuleList()
+
+        for n in range(num_blocks):
+            logging.debug('Prepare block %d', n)
+
+            channel_in_features = self.features.num_features if n == 0 else temporal_block_dimension
```
Comment (titu1994) [inline nemo/collections/asr/modules/audio_modules.py:558]: It's a lot of code config, could you add some comments about what these sections are ?

File: nemo/collections/asr/modules/audio_modules.py [hunk 8]
```diff
@@ -562,7 +1015,7 @@ def forward(
             as the input length.
         """
         # Temporal weighting: average power over channels, shape (B, F, N)
-        weight = torch.mean(power, dim=1)
+        weight = torch.mean(power, dim=-3)
```
Comment (titu1994) [inline nemo/collections/asr/modules/audio_modules.py:1018]: Any reason to use -3 rather than 1?

File: nemo/collections/asr/parts/submodules/multichannel_modules.py [hunk 9]
```diff
@@ -0,0 +1,778 @@
...
+    def __init__(
+        self,
+        permute_channels: bool = True,
+        num_channels_min: int = 1,
+        num_channels_max: Optional[int] = None,
+        rng: Optional[Callable] = None,
+    ):
+        super().__init__()
+
+        self._rng = random.Random() if rng is None else rng
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multichannel_modules.py:52]: Maybe some provision to set the seed ?

File: nemo/collections/asr/parts/submodules/multichannel_modules.py [hunk 10]
```diff
@@ -0,0 +1,778 @@
...
+
+        if self.permute_channels:
+            self._rng.shuffle(channels)
+
+        channels = channels[:num_channels_out]
+
+        return input[:, channels, :]
+
+
+class TransformAverageConcatenate(NeuralModule):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multichannel_modules.py:108]: These modules and their functions need independent unit tests

File: nemo/collections/asr/parts/submodules/multichannel_modules.py [hunk 11]
```diff
@@ -0,0 +1,778 @@
...
+        ref_subband_weighting: bool = False,
+        num_subbands: Optional[int] = None,
+        diag_reg: Optional[float] = 1e-6,
+        eps: float = 1e-8,
+    ):
+        if not HAVE_TORCHAUDIO:
+            logging.error('Could not import torchaudio. Some features might not work.')
+
+            raise ModuleNotFoundError(
+                "torchaudio is not installed but is necessary to instantiate a {self.__class__.__name__}"
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multichannel_modules.py:428]: Needs fstring here

File: nemo/collections/asr/parts/submodules/multichannel_modules.py [hunk 12]
```diff
@@ -0,0 +1,778 @@
...
+               C is the number of input channels and M is the number of output channels
+            psd_s: Covariance for the signal, shape (B, F, C, C)
+            psd_n: Covariance for the noise, shape (B, F, C, C)
+
+        Returns:
+            One-hot or soft reference channel, shape (B, M)
+        """
+        if self.subband_weighting:
+            # (B, F, M)
+            pow_s = torch.einsum('...jm,...jk,...km->...m', W.conj(), psd_s, W).abs()
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multichannel_modules.py:745]: Just a note if you ever intend to export to onnx in the future, it doesn't have great support of einsum

**Review by titu1994 — COMMENTED** [review]
Excellent PR, very well documented and written. Mostly good to merge, but Id like some unit tests for the multiple modules added in this pr before merge. 

Minor comments overall. 

I noted the debug statements, but you can ignore them. They seem useful.

**Review by titu1994 — DISMISSED** [review]

**Review by titu1994 — DISMISSED** [review]
LGTM

**Review by titu1994 — COMMENTED** [review]
Lets hold off merging until we can remove the pleasefixme's

**Review by titu1994 — APPROVED** [review]
Looks great, @Kipok and @GNroy please look into the pleasefixme tests when you get time.
